### PR TITLE
Fixes error handling in firebase_auth on Android and iOS

### PR DESCRIPTION
**Pull request by slightfoot**
This standardises the error responses from the platform implementations. All functions in `FirebaseAuth` and `FirebaseUser` are now documented with error codes.

Fixes: flutter/flutter#18312 and Fixes: flutter/flutter#20223

This PR is based on #678 so includes the changes there.

**Example:**
```
PlatformException(ERROR_USER_NOT_FOUND, There is no user record corresponding to this identifier. The user may have been deleted., null)
PlatformException(ERROR_WRONG_PASSWORD, The password is invalid or the user does not have a password., null)
PlatformException(ERROR_NETWORK_REQUEST_FAILED, Network error (such as timeout, interrupted connection or unreachable host) has occurred., null)
```

Other changes include:
* Added missing `linkWithTwitterCredential` method.
* Updated deprecated API `fetchProvidersForEmail` to `fetchSignInMethodsForEmail`.
* Added `unlinkCredential`.
* Updated Changelog.

Full error code list can be found here:
https://docs.google.com/spreadsheets/d/1FDn0rRRYjgXMwc_FIAxO7_cQh69q5VXgQt7Hmvyb1Sg/edit?usp=sharing

Closes: #768 and Closes: #697 - Merged into this one.
Closes: #703 and #814

### CODE REVIEW

**Comment by slightfoot** [comment]
@kroikie this is now ready for review and merge.

**Comment by slightfoot** [comment]
Another big change in this code is that now all the dart code of `firebase_auth` is now documented.

@kroikie I just noticed my PR overlaps several others noted here. I just fixed things as I saw they were missing when adding the error logging.

* #703 is included as `unlinkCredential` on the platform interface and `unlinkEmailAndPassword`/`unlinkGoogleCredential`/`unlinkFacebookCredential`/`unlinkTwitterCredential` on the `FirebaseUser` object. This matches the style of API we have in dart for signIn and link.
* #697 this is now defunct, all functionality is in this PR and has been documented also.
* #768 overlaps with #697 as it provides Twitter and Github login. I can merge this PR into this one and provide Github login feature with documentation if you like.

Would be nice to see #718 merged in the same version release as this.

**Comment by filleduchaos** [comment]
@slightfoot I don't mind you taking the changes from #768! :)

**Comment by escamoteur** [comment]
@eseidelGoogle could you check on this please? It's really a needed fix.

**Comment by eseidelGoogle** [comment]
@bparrishMines may also be able to help with reviews.

**Comment by bparrishMines** [comment]
HI @slightfoot,

Thanks for your contribution! Could you resolve the recent conflicts? I'll finish reviewing it as soon as you do.

**Comment by slightfoot** [comment]
@bparrishMines think the analyze queue broken. Its been waiting a few hours. I'll double check all the formatting is good tomorrow when the results come in. Other than that, this is ready to go.

**Comment by slightfoot** [comment]
@kroikie looks like the build server is borked. perhaps you can kick it. 😉

**Comment by bparrishMines** [comment]
cc @fkorotkov @gspencergoog Do you know why the analyze check is stuck here?

**Comment by HerrNiklasRaab** [comment]
When will this change be released?

**Comment by beardo01** [comment]
I support getting this merged too. Let's get it done please.

**Comment by kroikie** [comment]
@slightfoot Sorry about the delay on this PR. Would you be able to rebase this on master then I'd be happy to merge.

**Comment by manhnhhdvn** [comment]
@bparrishMines , @kroikie

This PR is wonderful, but if it seems time consuming, could you review #703  and #814 first?

#703 is the ```unlink``` part of this PR.
#814 is the ```fetchSignInMethodsForEmail``` part of this PR.

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 1]
```diff
@@ -607,4 +672,32 @@ public void onComplete(@NonNull Task<ProviderQueryResult> task) {
...
+
+  private void reportException(Result result, @Nullable Exception exception) {
+    if (exception != null) {
+      if (exception instanceof FirebaseAuthException) {
+        final FirebaseAuthException authException = (FirebaseAuthException) exception;
+        result.error(authException.getErrorCode(), exception.getMessage(), null);
+      } else if (exception instanceof FirebaseApiNotAvailableException) {
+        result.error(
+            "ERROR_API_NOT_AVAILABLE",
+            "The API that you are calling is not available on devices without Google Play Services.",
```
[the remark this answers is not in the document]
Comment (slightfoot) [reply packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: Changed.

File: packages/firebase_auth/lib/firebase_auth.dart [hunk 2]
```diff
@@ -74,28 +75,54 @@ class FirebaseUser extends UserInfo {
 
   /// Obtains the id token for the current user, forcing a [refresh] if desired.
   ///
+  /// Useful when authenticating against your own backend. Use our server
+  /// SDKs or follow the official documentation to securely verify the
+  /// integrity and validity of this token.
+  ///
   /// Completes with an error if the user is signed out.
+  ///
```
[the remark this answers is not in the document]
Comment (slightfoot) [reply packages/firebase_auth/lib/firebase_auth.dart]: Thanks, changed these also.

File: packages/firebase_auth/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,12 @@
+## 0.6.1
+
+* `FirebaseAuth` and `FirebaseUser` are now fully documented.
+* `PlatformExceptions` now report error codes as stated in docs.
+* Added support for `linkWithTwitterCredential` in `FirebaseAuth`.
+* Added support for GitHub to `FirebaseAuth`.
+* **Breaking Change** Renamed `fetchProvidersForEmail` to its new name `fetchSignInMethodsForEmail`.
```
[the remark this answers is not in the document]
Comment (slightfoot) [reply packages/firebase_auth/CHANGELOG.md]: Updated.

**Comment by slightfoot** [comment]
@bparrishMines @kroikie It's ready to merge.

**Comment by kroikie** [comment]
@slightfoot sorry for the churn here, could you rebase then should be good to merge.

**Comment by slightfoot** [comment]
Rebased.

**Comment by slightfoot** [comment]
@kroikie I've rebased again, and merged whilst I can, but leave the release to you.

**Comment by xqwzts** [comment]
@slightfoot Thanks for the long-awaited and hugely useful commit.

An official release would be appreciated as this fixes some critical bugs wrt error handling.

Renaming `fetchProvidersForEmail` to `fetchSignInMethodsForEmail` is a breaking change though, surely it warrants a mention in the changelog and a version bump?

**Comment by slightfoot** [comment]
@xqwzts I was asked to remove my "breaking change" mention from the CHANGELOG and the @Deprecated function I had because we have not yet reached stable API version, or something.

**Comment by HerrNiklasRaab** [comment]
How is the status? When will this be in beta?

**Comment by Daibaku0615** [comment]
@slightfoot Do you know when this will be available?

**Comment by georgeherby** [comment]
If you want these changes before an official release but without the stuff in the proposed 0.7.0 in the change log
just add this to your yaml
```
  firebase_auth:
    git:
      url: git://github.com/flutter/plugins.git
      ref: e96fb0c350328ff77baae758b6dc70797117ac98
      path: packages/firebase_auth
```
The commit is of the stuff in this PR which in the changelog is 0.6.7 - just get the most recent commit if you want the stuff in 0.7.0 but that is nothing to do with this PR

**Comment by slightfoot** [comment]
@kroikie is it possible to push this out. A lot of people are still waiting for this. `0.6.7` never got released. we are still on `0.6.6`.

**Comment by sumitrvi** [comment]
Hey i'm also facing google signin issue in ios
Unhandled Exception: PlatformException(ERROR_INTERNAL_ERROR, An internal error has occurred, print and inspect the error details for more information., null)
#0 StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:569:7)
#1 MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:321:33)

#2 MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:349:48)
#3 MethodChannelFirebaseAuth.signInWithCredential (package:firebase_auth_platform_interface/src/method_channel_firebase_auth.dart:161:23)
#4 FirebaseAuth.signInWithCredential (package:firebase_auth/src/firebase_auth.dart:208:10)
#5 _AsyncAwaitCompleter.start (dart:async-patch/async_patch.dart:45:6)
#6 FirebaseAuth.signInWithCredential (package:firebase_auth/src/firebase_auth.dart:205:42)
#7 _SigninScreenState.socialLogin (package:down_beatflutterproject/screens/signinscreen.<…>

**Pull request merged**